crates/host/src/wasmbus/providers/messaging_nats.rs

```diff
@@ -113,11 +175,69 @@ impl wasmcloud_provider_sdk::Provider for Provider {
             ..
         }: LinkConfig<'_>,
     ) -> anyhow::Result<()> {
-        use wrpc::exports::wasmcloud::messaging0_2_0::handler::Handler as _;
-
         let (nats, config) = self.connect(config).await?;
         let mut tasks = JoinSet::new();
         let target_id: Arc<str> = Arc::from(target_id);
+        for ConsumerConfig {
+            stream,
+            consumer,
+            max_messages,
+            max_bytes,
+        } in config.consumers
+        {
+            let js = jetstream::new(nats.clone());
+            let stream = js
+                .get_stream(stream)
+                .await
+                .context("failed to get stream")?;
+            let consumer = stream
+                .get_consumer(&consumer)
+                .await
+                .map_err(|err| anyhow!(err).context("failed to get consumer"))?;
+            let sub = consumer.batch();
+            let sub = if let Some(max_messages) = max_messages {
+                sub.max_messages(max_messages)
+            } else {
+                sub
+            };
+            let sub = if let Some(max_bytes) = max_bytes {
+                sub.max_bytes(max_bytes)
+            } else {
+                sub
+            };
+            let mut sub = sub.messages().await.context("failed to subscribe")?;
+
+            let components = Arc::clone(&self.components);
+            let lattice_id = Arc::clone(&self.lattice_id);
+            let host_id = Arc::clone(&self.host_id);
+            let target_id = Arc::clone(&target_id);
+            tasks.spawn(async move {
```

Comment:
When the component stops, or is unlinked from the provider, we should probably cancel this task so that we don't keep handling messages. Perhaps we can store it as a part of the messaging provider struct and then remove it if the link is deleted?

Reply:
That's already what happens - the task is cancelled when link is deleted https://github.com/wasmCloud/wasmCloud/blob/9926109f0e3825cd3e28e53469fa6e23d41dbba7/crates/host/src/wasmbus/providers/messaging_nats.rs#L218-L222

Component stops are unrelated here - the link can be established before it even starts running

Reply:
My brain did not parse `tasks.spawn` as going into the JoinSet, my mistake 😄